Fabric.Authorization.Persistence.SqlServer/Stores/SqlServerPermissionStore.cs

```diff
@@ -0,0 +1,206 @@
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == model.Id
+                    && !p.IsDeleted);
+
+            if (permission == null)
+            {
+                throw new NotFoundException<User>($"Could not find {typeof(Permission).Name} entity with ID {model.Id}");
+            }
+
+            permission.IsDeleted = true;
+            await _authorizationDbContext.SaveChangesAsync();
+        }
+
+        public async Task Update(Permission model)
+        {
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == model.Id
+                    && !p.IsDeleted);
+
+            if (permission == null)
+            {
+                throw new NotFoundException<User>($"Could not find {typeof(Permission).Name} entity with ID {model.Id}");
+            }
+
+            model.ToEntity(permission);
+            _authorizationDbContext.Permissions.Update(permission);
+            await _authorizationDbContext.SaveChangesAsync();
+        }
+
+        public async Task<bool> Exists(Guid id)
+        {
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == id
+                    && !p.IsDeleted);
+
+            return permission != null;
+        }
+
+        public Task<IEnumerable<Permission>> GetPermissions(string grain, string securableItem = null,
+            string permissionName = null)
+        {
+            var permissions = _authorizationDbContext.Permissions
+                .Include(p => p.SecurableItem)
+                .Where(p => string.Equals(p.Grain, grain, StringComparison.OrdinalIgnoreCase));
+
+            if (!string.IsNullOrWhiteSpace(securableItem))
+            {
+                permissions = permissions.Where(p => string.Equals(p.SecurableItem.Name, securableItem));
+            }
+
+            if (!string.IsNullOrWhiteSpace(permissionName))
+            {
+                permissions = permissions.Where(p => string.Equals(p.Name, securableItem));
+            }
+
+            return Task.FromResult(permissions.Select(p => p.ToModel()).AsEnumerable());
+        }
+
+        public async Task AddOrUpdateGranularPermission(GranularPermission granularPermission)
+        {
+            var idParts = SplitUserId(granularPermission.Id);
+
+            var user = await _authorizationDbContext.Users
+                .Include(u => u.UserPermissions)
+                .ThenInclude(up => up.Permission)
+                .SingleOrDefaultAsync(u => u.IdentityProvider.Equals(idParts[0], StringComparison.OrdinalIgnoreCase)
+                                           && u.SubjectId.Equals(idParts.Length > 1 ? idParts[1] : idParts[0], StringComparison.OrdinalIgnoreCase)
+                                           && !u.IsDeleted);
+
```

Comment:
i think there is a null check missing for user

Reply:
Good catch - I'll fix this.